From 97d2681f9e3bda782fe6c544724cb5d0e9b2b0d1 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:25:06 -0400
Subject: [tools] perf/trace: Beautify pipe2 'flags' arg

Message-id: <1408462094-14194-138-git-send-email-jolsa@redhat.com>
Patchwork-id: 88156
O-Subject: [PATCH RHEL7.1 BZ1131394 137/325] perf trace: Beautify pipe2 'flags' arg
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 46cce19ba8cc8231d3c949c128e4bdaf420228a2
Author: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Mon Sep 23 12:52:04 2013 -0300

description
===========
 4.234 (0.005 ms): fetchmail/3224 pipe2(fildes: 0x7fffc72bcee0, flags: CLOEXEC) = 0
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-trace.c b/tools/perf/builtin-trace.c
index 245534f..adc7cd0 100644
--- a/tools/perf/builtin-trace.c
+++ b/tools/perf/builtin-trace.c
@@ -517,6 +517,29 @@ static size_t syscall_arg__scnprintf_eventfd_flags(char *bf, size_t size,
 
 #define SCA_EFD_FLAGS syscall_arg__scnprintf_eventfd_flags
 
+static size_t syscall_arg__scnprintf_pipe_flags(char *bf, size_t size,
+      struct syscall_arg *arg)
+{
+ int printed = 0, flags = arg->val;
+
+#define P_FLAG(n) \
+ if (flags & O_##n) { \
+  printed += scnprintf(bf + printed, size - printed, "%s%s", printed ? "|" : "", #n); \
+  flags &= ~O_##n; \
+ }
+
+ P_FLAG(CLOEXEC);
+ P_FLAG(NONBLOCK);
+#undef P_FLAG
+
+ if (flags)
+  printed += scnprintf(bf + printed, size - printed, "%s%#x", printed ? "|" : "", flags);
+
+ return printed;
+}
+
+#define SCA_PIPE_FLAGS syscall_arg__scnprintf_pipe_flags
+
 static size_t syscall_arg__scnprintf_signum(char *bf, size_t size, struct syscall_arg *arg)
 {
  int sig = arg->val;
@@ -620,6 +643,8 @@ static struct syscall_fmt {
    .arg_scnprintf = { [2] = SCA_OPEN_FLAGS, /* flags */ }, },
  { .name     = "openat",     .errmsg = true,
    .arg_scnprintf = { [2] = SCA_OPEN_FLAGS, /* flags */ }, },
+ { .name     = "pipe2",     .errmsg = true,
+   .arg_scnprintf = { [1] = SCA_PIPE_FLAGS, /* flags */ }, },
  { .name     = "poll",     .errmsg = true, .timeout = true, },
  { .name     = "ppoll",     .errmsg = true, .timeout = true, },
  { .name     = "pread",     .errmsg = true, .alias = "pread64", },
-- 
1.7.1